### PR TITLE
Improve mince and add method for non-uniform partition of IntervalBox

### LINKED ISSUE #444: Generalize IntervalBox split for non-uniform partitions
Currently, `mince` for interval boxes receives an integer `n` to divide the box n times along each dimension,

https://github.com/JuliaIntervals/IntervalArithmetic.jl/blob/b70940dabdbefd8eb4be32d7bde6df72174dd3f7/src/multidim/intervalbox.jl#L124

Sometimes it is useful to have a function that accepts a non-uniform partition, say `[2, 3]` to split the first dimension two times and the second dimension 3 times, obtaning 6 boxes. (This is what `split(::AbstractHyperrectangle, block::AbstractVector)` does.)

### PR DESCRIPTION
**Pull request by lbenet**
Closes #444

### CODE REVIEW

**Comment by lbenet** [comment]
Some benchmarking:

- Current master:
```
julia> @benchmark mince(ii, ndiv) setup=(ii=-1 .. 1, ndiv=100_000_000)
BenchmarkTools.Trial: 
  memory estimate:  152.59 MiB
  allocs estimate:  2
  --------------
  minimum time:     45.498 ms (0.00% GC)
  median time:      64.020 ms (26.33% GC)
  mean time:        58.819 ms (17.82% GC)
  maximum time:     120.155 ms (58.98% GC)
  --------------
  samples:          85
  evals/sample:     1
```

- This PR
```
julia> @benchmark mince(ii, ndiv) setup=(ii=-1 .. 1, ndiv=100_000_000)
BenchmarkTools.Trial: 
  memory estimate:  152.59 MiB
  allocs estimate:  2
  --------------
  minimum time:     41.535 ms (0.00% GC)
  median time:      53.344 ms (16.84% GC)
  mean time:        54.540 ms (19.06% GC)
  maximum time:     117.183 ms (60.92% GC)
  --------------
  samples:          92
  evals/sample:     1
```

File: src/intervals/arithmetic.jl [hunk 1]
```diff
@@ -577,5 +577,5 @@ as a vector.
 """
 function mince(x::Interval, n)
     nodes = range(x.lo, x.hi, length = n+1)
-    return [Interval(nodes[i], nodes[i+1]) for i in 1:length(nodes)-1]
+    return @. Interval(nodes[1:n], nodes[2:n+1])
```
Comment (Kolaru) [inline src/intervals/arithmetic.jl:580]:
```suggestion
    return Interval.(nodes[1:n], nodes[2:n+1])
```
Should be equivalent and avoid the (rather uncommon) macro.

Comment (lbenet) [reply src/intervals/arithmetic.jl]: Thanks for the suggestion!

File: src/multidim/intervalbox.jl [hunk 2]
```diff
@@ -138,6 +138,26 @@ as a vector.
     end
 end
 
+@generated function mince(x::IntervalBox{N,T}, tup::NTuple{N,Int}) where {N,T}
+    quote
+        n = maximum(tup)
+        nodes_matrix = Array{Interval{T},2}(undef, n, N)
+        fill!(nodes_matrix, emptyinterval(T))
```
Comment (Kolaru) [inline src/multidim/intervalbox.jl:145]:
```suggestion
        nodes_matrix = fill(emptyinterval(T), n, N)
```
should be equivalent.

File: src/multidim/intervalbox.jl [hunk 3]
```diff
@@ -138,6 +138,26 @@ as a vector.
...
+        for i in 1:N
+            nodes_matrix[1:tup[i],i] .= mince(x[i], tup[i])
+        end
+
+        nodes = IntervalBox{$N,T}[]
+        Base.Cartesian.@nloops $N i _->(1:n) begin
```
Comment (Kolaru) [inline src/multidim/intervalbox.jl:151]: Does this part requires carefull review? I am not familiar with the macros used, but I could delve into it if needed.

File: src/multidim/intervalbox.jl (hunk 2 above)
Comment (lbenet) [reply src/multidim/intervalbox.jl]: Thanks! Done!

File: src/multidim/intervalbox.jl (hunk 3 above)
Comment (lbenet) [reply src/multidim/intervalbox.jl:150]: Thanks for your review @Kolaru and also for offering getting into the details of this part of the code. 

Here, I simply used the original method, and adapt it to obtain the desired behavior; maybe there are better ways, but for the time being, this is good enough. The original code was inspired from [this section of the manual](https://docs.julialang.org/en/v1/devdocs/cartesian/#Base.Cartesian), which admittedly I have mostly forgotten.

Comment (Kolaru) [reply src/multidim/intervalbox.jl:150]: That sounded like a fun multidimensional challenge so I spent some time on it and I believe the following is equivalent:
```julia
function mince(x::IntervalBox{N,T}, ncuts::NTuple{N,Int}) where {N,T}
    minced_intervals = [mince(x[i], ncuts[i]) for i in 1:N]
    minced_boxes = Vector{IntervalBox{N,T}}(undef, prod(ncuts))

    for (k, cut_indices) in enumerate(CartesianIndices(ncuts))
        ibox = IntervalBox([minced_intervals[i][cut_indices[i]] for i in 1:N])
        minced_boxes[k] = ibox
    end
    return minced_boxes
end
```

Comment (lbenet) [reply src/multidim/intervalbox.jl:150]: Thanks @Kolaru for the suggestion! I am also changing the other method (avoiding then the `@generated`). I would like to [include you as coauthor](https://docs.github.com/en/github/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors#required-co-author-information) of this commit; is it ok with you? (I already have the commit ready, but prefer to wait for your reply before pushing it.)

Comment (Kolaru) [reply src/multidim/intervalbox.jl:150]: Sure let me know if it needs any input from my side.

Comment (lbenet) [reply src/multidim/intervalbox.jl]: I just pushed it, so you can have a look

**Comment by Kolaru** [comment]
All look good, thanks.

**Comment by lbenet** [comment]
Thanks a lot!

**Pull request merged**